### PR TITLE
docs: note transposed weight initialisations

### LINKED ISSUE #129834: Initialisation docs should mention that `weight` matrices are transposed
### 📚 The doc issue

The docs do not mention that weight matrices are assumed to be initialised as: `w.shape = [fan_out, fan_in]`, because Pytorch linear layers do: `h = x @ w.T + b`

Many people initialising weights themselves would do: `w.shape = [fan_in, fan_out]` and `h = x @ w + b`; leading to subtle bugs — unless they compare the variance of the weight matrix with the variance of the Pytorch linear layers, and look into the source code, they may not catch that the `[fan_in, fan_out]` used by the initialization function is not what they expect.

### Suggest a potential alternative/fix

The initialisation functions should clearly state that this is the case.

--- 

### Relevant links / examples of confusion caused

- https://stackoverflow.com/questions/53462493/shape-of-pytorch-model-parameter-is-inconsistent-with-how-its-defined-in-the-mo/
- https://stackoverflow.com/questions/53465608/pytorch-shape-of-nn-linear-weights
- https://github.com/pytorch/pytorch/issues/2159

cc @svekars @brycebortree @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### PR DESCRIPTION
**Pull request by sradc**
Fixes #129834

### CODE REVIEW

File: torch/nn/init.py [hunk 1]
```diff
@@ -388,6 +388,13 @@ def xavier_uniform_(
     Examples:
         >>> w = torch.empty(3, 5)
         >>> nn.init.xavier_uniform_(w, gain=nn.init.calculate_gain('relu'))
+
+    Note:
+        Be aware that `fan_in` and `fan_out` are calculated assuming that the weight
+        matrix is used in a transposed manner (i.e., `x @ w.T` in `Linear` layers).
+        For `Linear` layers, `w.shape = [fan_out, fan_in]`. This is important for
+        correct initialization. If you plan to use `x @ w`, pass in
+        `nn.init.xavier_uniform_(w.T, ...)`, where `w.shape = [fan_in, fan_out]`.
```
Comment (mikaylagawarecki) [inline torch/nn/init.py:397]: You can view the rendered text over here https://github.com/pytorch/pytorch/pull/130122#issuecomment-2209375977 in "view Python docs built from this PR"
Screenshot 2024-07-10 at 3 38 38 PM

Use double backticks for nice rendering
```suggestion
        Be aware that ``fan_in`` and ``fan_out`` are calculated assuming that the weight
        matrix is used in a transposed manner (i.e., ``x @ w.T`` in ``Linear`` layers).
        For ``Linear`` layers, ``w.shape = [fan_out, fan_in]``. This is important for
        correct initialization. If you plan to use ``x @ w``, pass in
        ``nn.init.xavier_uniform_(w.T, ...)``, where ``w.shape = [fan_in, fan_out]``.
```

similarly for the rest

**Comment by mikaylagawarecki** [comment]
Also, please sign cla!

**Comment by sradc** [comment]
Thank you @mikaylagawarecki! I've updated the backticks and signed the CLA

**Review by mikaylagawarecki — APPROVED** [review]
Thanks!

**Pull request closed without merging**